src/rgw/driver/rados/rgw_bucket.cc

```diff
@@ -1445,10 +1446,35 @@ int RGWBucketAdminOp::remove_bucket(rgw::sal::Driver* driver, RGWBucketAdminOpSt
   if (ret < 0)
     return ret;
 
+  // check if the bucket is owned by my zonegroup, if not, refuse to remove it
+  if (bucket->get_info().zonegroup != driver->get_zone()->get_zonegroup().get_id()) {
+    ldpp_dout(dpp, 0) << "ERROR: The bucket's zonegroup does not match. "
+                      << "Removal is not allowed. Please execute this operation "
+                      << "on the bucket's zonegroup: " << bucket->get_info().zonegroup << dendl;
+    return -ERR_PERMANENT_REDIRECT;
+  }
+
   if (bypass_gc)
     ret = bucket->remove_bypass_gc(op_state.get_max_aio(), keep_index_consistent, y, dpp);
   else
     ret = bucket->remove(dpp, op_state.will_delete_children(), y);
+  if (ret < 0)
+    return ret;
+
+  // forward to master zonegroup
+  const std::string delpath = "/" + bucket->get_name();
```

Comment:
i don't think this will do the right thing for tenanted buckets

Reply:
right, thanks for noticing this. Do you happen to know any helper function perhaps around `rgw_bucket` that can generate the right URL (I assume `tenant + ':' + name`)? The only near I've found was `rgw_bucket::get_key()` which use `/` as the `tenant_delim` and is also including bucket_id which I guess it's irrelevant.

Reply:
https://github.com/ceph/ceph/pull/60848 recently deprecated
> S3 API support for cross-tenant names such as `Bucket='tenant:bucketname'`

so we shouldn't expect that to work for s3. there's the [DELETE /admin/bucket](https://github.com/ceph/ceph/blob/093e0de0/src/rgw/driver/rados/rgw_rest_bucket.cc#L223) api which the multisite system user could access, but we'd need to teach it to take a `tenant` param

Reply:
thanks for the update. I changed it to the admin API and teached [RGWOp_Bucket_Remove](https://github.com/ceph/ceph/pull/60227/commits/a7e6510535ded28bf9571d095863e12bb19422b5) to take `tenant` param.